### PR TITLE
fix(maven): assemble XML text split by entity references

### LINKED ISSUE #394: Maven XML values containing entity references are truncated
`Event::GeneralRef` is not handled in the Maven XML paths, so any value containing an XML entity reference is truncated to its last text fragment.

quick-xml emits `&amp;` (and friends) as a separate `Event::GeneralRef` between two `Event::Text` events. `parsers/maven.rs` and `registries/maven_central.rs` only match `Event::Text`, and each fragment *replaces* the accumulator instead of appending to it, so only the fragment after the last entity survives.

## Reproduction

```rust
let pom = r#"<?xml version="1.0"?>
<project>
  <dependencies>
    <dependency>
      <groupId>org.a&amp;b</groupId>
      <artifactId>art&amp;fact</artifactId>
      <version>3.1&amp;4</version>
    </dependency>
  </dependencies>
</project>"#;
let deps = MavenParser::new().parse(pom);
// got:      name = "b:fact", version = "4"
// expected: name = "org.a&b:art&fact", version = "3.1&4"
```

The reported spans are wrong the same way — they cover only the trailing fragment.

## Impact

Maven coordinates are restricted to `[A-Za-z0-9_.-]`, so `groupId`/`artifactId`/`version` are not reachable in practice. The reachable case is `parse_pom_metadata` in `registries/maven_central.rs`: `<description>`, `<url>`, `<scm><url>` and `<licenses><license><name>` routinely contain `&amp;` (query strings, "X & Y" license names). Those get truncated in hover output.

`extract_properties` (`parsers/maven.rs`) has the same shape and would truncate a property value containing an entity.

## Notes

Pre-existing, not introduced by the quick-xml 0.42 bump in #393 — verified by running the reproduction above against `main` at a57b83c (quick-xml 0.41), which produces byte-identical output.

## Fix sketch

Accumulate `Event::Text` and `Event::GeneralRef` into one buffer per element, resolve the reference with `BytesRef::resolve_char_ref()` / `escape::resolve_xml_entity()`, and flush on `Event::End`. Track the span from the first fragment's start to the last fragment's end. Needs regression coverage for escaped values in both the dependency path and the metadata path.

### PR DESCRIPTION
**Pull request by mpiton**
## Summary

`Event::GeneralRef` was not handled in the Maven XML paths, so any value containing an XML entity reference was truncated to a single fragment. quick-xml reports every `&…;` inside element content as a standalone `Event::GeneralRef` between two `Event::Text` events, and the four Maven loops matched `Event::Text` only, each fragment *replacing* the accumulator instead of appending to it.

`extract_dependencies` kept the last fragment, `extract_properties` the first (its `!out.contains_key` guard). Maven coordinates are restricted to `[A-Za-z0-9_.-]`, so the reachable damage was in the pom metadata shown on hover — descriptions, URLs with query strings, "X & Y" license names — and in property values.

A second cause, not in the issue: `trim_text(true)` trims each fragment on its own, so simply concatenating would have turned `GPL &amp; Classpath` into `GPL&Classpath`.

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Refactoring (no functional changes)

## Related Issue

Closes #394

## Changes Made

- `utils::push_xml_ref` resolves a `BytesRef`: numeric character references via `resolve_char_ref`, the five predefined XML entities via `escape::resolve_xml_entity`. Anything else — a DTD-declared entity, a typo — is written back verbatim as `&name;` instead of being dropped.
- Element text is accumulated across `Text` and `GeneralRef` fragments and flushed on `End`, in `extract_properties`, `extract_dependencies` (`parsers/maven.rs`), `parse_metadata_xml` and `parse_pom_metadata` (`registries/maven_central.rs`).
- The three loops that ran with `trim_text(true)` move to `trim_text(false)` plus a manual trim of the assembled value. Flushing is skipped when the trimmed value is empty, which keeps the existing behaviour of empty and whitespace-only elements yielding no value.
- Spans were wrong the same way. `trimmed_span` derived its start from `buffer_position() - raw_len`, which only holds when a value arrives as a single event; it now takes the reader position captured *before* `read_event()` for the first fragment through the position after the last. `<version>3.1&amp;4</version>` reports the whole run instead of the trailing `4`.
- CHANGELOG entry under `[Unreleased] > Fixed`.

## Testing

- [x] `cargo test` passes — 923 passed, 0 failed
- [x] `cargo clippy -- -D warnings` passes
- [x] `cargo fmt --check` passes
- [ ] Tested manually in Zed

9 regression tests. The issue's reproduction is `test_parse_dependency_with_entity_references` verbatim. The rest cover spans, numeric character references (`&#45;`), an unresolvable entity kept as written, spaces around `&amp;`, property substitution, `<description>`/`<url>`/`<scm><url>`/multiple licenses, `maven-metadata.xml`, and empty elements still yielding `None`.

## Checklist

- [x] My code follows the project's style guidelines
- [x] I have updated documentation if needed
- [x] I have added tests for new functionality
- [x] All new and existing tests pass

## Out of scope

- `csharp.rs` is untouched. It already handles `Event::GeneralRef` and marks the version fragmented via `mark_fragmented_version`, dropping the dependency rather than reporting a truncated version — deliberate and correct.
- CDATA is still ignored in the Maven paths: `<description><![CDATA[Fast & small]]></description>` yields `None`. Same class of lost value, but a different gap (no `Event::CData` arm at all) and not what #394 reports.

## Follow-ups worth considering

- The four text-assembly loops remain copy-pasted; only `push_xml_ref` is shared. A common accumulator would stop the next copy from repeating this.
- There are 8 fuzz targets and none for Maven/XML. Fuzzing pom.xml would have caught this.

---
## Summary by cubic
Fixes Maven XML parsing so element values containing entity references or CDATA are no longer truncated or lost (#394), and skips dependencies whose coordinates can't be located exactly instead of misreporting them.

**Bug Fixes**

- Element text is assembled across `Text`, `CData`, and `GeneralRef` events; `push_xml_ref` resolves numeric and predefined XML entities and preserves unresolvable ones verbatim.
- Values are trimmed once after assembly, keeping spaces around `&amp;`, and spans now cover the whole raw value.
- Only direct children of `<dependency>` set `groupId`, `artifactId`, and `version`, so an `<exclusion>` can't overwrite them.
- A coordinate that is unexpandable, split across lines, or interrupted by a comment is skipped, because it has no valid single-line span for the update quick-fix.
- A property or `maven-metadata.xml` version holding an unresolvable entity is dropped, so `${key}` stays unsubstituted and the literal `&name;` never reaches the registry.

<sup>Written for commit 8a6dd546a48f4092d0fc86fa4926450dd87ba960. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/mpiton/zed-depsy/pull/395?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved Maven XML parsing for CDATA, entity references, whitespace, and empty values.
  * Preserved unknown entity references in free text while safely omitting unresolved properties and metadata versions.
  * Corrected dependency and exclusion coordinate extraction.
  * Skipped dependencies with unresolved, fragmented, multiline, or unreliable coordinates and versions.
  * Improved source-range accuracy for interrupted XML values.

* **Tests**
  * Added coverage for malformed XML and source-span safety across Maven edge cases.

### CODE REVIEW

File: depsy-lsp/src/parsers/maven.rs [hunk 1]
```diff
@@ -294,24 +339,14 @@ fn extract_dependencies(content: &str, properties: &HashMap<String, String>) ->
...
-                    _ => {}
-                }
+                text.push_str(&e);
+                let start = text_span.map_or(event_start, |(s, _)| s);
+                text_span = Some((start, reader.buffer_position() as usize));
+            }
+            Ok(Event::GeneralRef(e)) if in_dependency => {
+                push_xml_ref(&mut text, &e);
+                let start = text_span.map_or(event_start, |(s, _)| s);
+                text_span = Some((start, reader.buffer_position() as usize));
```
Comment (mpiton) [inline depsy-lsp/src/parsers/maven.rs]: > [!IMPORTANT]
> **🟡 Medium · 🐛 Correctness** — A comment or CDATA inside an element is swallowed into the value and its span

`Event::Comment`, `Event::CData` and `Event::PI` fall to the catch-all `_ => {}`, so they neither contribute to `text` nor reset it, while `text_span` keeps extending its end past them. Verified on this branch: `<version>1.0<!-- c -->2.0</version>` yields version `1.02.0` with `version_span` covering `1.0<!-- c -->2.0` (16 bytes), and `<version>1.<![CDATA[0]]>5</version>` yields `1.5` over `1.<![CDATA[0]]>5`. Before the PR the last fragment won, so both reported `2.0` / `5` over their own 3 and 1 bytes. The span is what `code_actions.rs` writes the new version into, so accepting the Update quick-fix on such a declaration deletes the comment or the CDATA wrapper.

**Fix:** Reset `text` and `text_span` on `Event::Comment` / `Event::CData` / `Event::PI` inside a dependency the way the `Start` arm does — or, for CDATA, append it like `Text` and extend the span, which also closes the CDATA gap listed as out of scope in the PR body. The same three arms are missing from the other three loops.

<details>
<summary>🔎 Details</summary>

- **Why it belongs to this PR:** `text_span` and the multi-fragment accumulator are introduced by this diff; before it the span came from a single `Text` event and could not stretch across a comment.

</details>

<sub>🤖 <b>mpiton-review</b> · product-logic · confidence 0.95</sub>

File: depsy-lsp/src/parsers/maven.rs [hunk 2]
```diff
@@ -210,6 +229,32 @@ fn extract_dependencies(content: &str, properties: &HashMap<String, String>) ->
...
+                                    text_span.map(|(s, e_)| trimmed_span(bytes, s, e_));
+                                cur_artifact = Some(value.to_string());
+                            }
+                            Some("version") => {
+                                cur_version_span =
+                                    text_span.map(|(s, e_)| trimmed_span(bytes, s, e_));
+                                cur_version = Some(value.to_string());
+                            }
+                            Some("scope") => cur_scope = Some(value.to_string()),
+                            Some("optional") => cur_optional = value == "true",
```
Comment (mpiton) [inline depsy-lsp/src/parsers/maven.rs]: > [!IMPORTANT]
> **🟡 Medium · 🐛 Correctness** — `<exclusions>` children overwrite the dependency's own groupId and artifactId

This dispatch matches on `current_tag` alone, with no check that the element is a direct child of `<dependency>`, so `<groupId>`/`<artifactId>` inside an `<exclusion>` overwrite `cur_group`, `cur_artifact` and `cur_artifact_span`. A pom declaring `real:art` and excluding `bad:badart` parses to `name = "bad:badart"` with `name_span` on the exclusion's line and the real `<version>` span, so hover, diagnostics, the registry lookup and the Update quick-fix all point at the excluded artifact. Not a regression — HEAD~1 does the same on the `Text` arm — but the dispatch moves onto these lines in this PR and the suite still has no `<exclusions>` coverage.

**Fix:** Track the depth at which `in_dependency` was set (both `Start` and `End` already run for every element) and apply this `match` only when the `End` is exactly one level below it. While you are here, `e.local_name()` is already recomputed 20 lines down — dispatching on it instead of `current_tag` removes the coupling with the `current_tag = None` reset 100 lines away. Add a regression test with an `<exclusions>` block asserting `deps[0].name == "real:art"`.

<details>
<summary>🔎 Details</summary>

- **Why it belongs to this PR:** the groupId/artifactId/version dispatch is moved from the `Text` arm onto these added lines by this diff, so the missing depth guard is now a one-line addition here.

</details>

<sub>🤖 <b>mpiton-review</b> · product-logic · confidence 0.95</sub>

File: depsy-lsp/src/parsers/maven.rs [hunk 3]
```diff
@@ -210,6 +229,32 @@ fn extract_dependencies(content: &str, properties: &HashMap<String, String>) ->
...
+                            Some("artifactId") => {
+                                cur_artifact_span =
+                                    text_span.map(|(s, e_)| trimmed_span(bytes, s, e_));
+                                cur_artifact = Some(value.to_string());
+                            }
+                            Some("version") => {
+                                cur_version_span =
+                                    text_span.map(|(s, e_)| trimmed_span(bytes, s, e_));
+                                cur_version = Some(value.to_string());
+                            }
```
Comment (mpiton) [inline depsy-lsp/src/parsers/maven.rs]: > [!IMPORTANT]
> **🟡 Medium · 🐛 Correctness** — An unresolvable entity now emits a dependency whose version is the literal `&name;`

The verbatim fallback in `push_xml_ref` is right for free text, but it changes what reaches the coordinate fields. A pom with `<!ENTITY ver "1.0">` and `<version>&ver;</version>` produced no `Event::Text` at all before, so `cur_version` stayed `None` and the dependency was skipped; on this branch it parses to `g:a` at version `&ver;`, verified against the head commit. `substitute` leaves it alone (no `${`), so the phantom version reaches the diagnostic and the Maven Central lookup. `<artifactId>&art;</artifactId>` likewise builds the request path from an unresolved entity.

**Fix:** Keep the verbatim fallback for `<description>`, `<url>` and licence names, and in `extract_dependencies` treat a reference `push_xml_ref` could not resolve as "value not understood": set a flag when it falls through and skip the commit for `groupId`/`artifactId`/`version`, so the dependency is dropped as before. Whichever way it goes, the module doc's `Unsupported in this MVP` list is the place to record it.

<details>
<summary>🔎 Details</summary>

- **Why it belongs to this PR:** the fallback is added by this diff at `utils.rs:117-119`, and it is what makes a previously-skipped dependency reachable.

</details>

<sub>🤖 <b>mpiton-review</b> · quality · confidence 0.9</sub>

File: depsy-lsp/src/parsers/maven.rs [hunk 4]
```diff
@@ -210,6 +229,32 @@ fn extract_dependencies(content: &str, properties: &HashMap<String, String>) ->
                 current_tag = Some(name);
             }
             Ok(Event::End(e)) => {
+                // Commit the element text before `</dependency>` clears the
+                // per-dependency state below.
+                if in_dependency {
+                    let value = text.trim();
+                    if !value.is_empty() {
```
Comment (mpiton) [inline depsy-lsp/src/parsers/maven.rs:277]: > [!NOTE]
> **🔵 Low · 🎯 Product** — A whitespace-only <version> now drops the dependency instead of reporting it

`!value.is_empty()` skips the whole `match`, so `cur_version_span` stays `None` and the `if let (Some(g), Some(a), Some((vs, ve)))` check below drops the dependency. Verified: `<version>   </version>` gives one `Dependency { version: "", .. }` at HEAD~1 and zero on this branch, so the declaration disappears from diagnostics, inlay hints and the document link. `<version></version>` was already skipped on both sides — only the whitespace-only form changes.

**Fix:** Looks deliberate and consistent with the `Skip dependencies that lack a <version>` policy, but it is a behaviour change outside what the CHANGELOG entry claims and no test pins it. Add a case next to `test_parse_dependency_without_version_is_skipped` asserting `parse` returns no dependency for a whitespace-only `<version>`, and mention it in the CHANGELOG entry.

<details>
<summary>🔎 Details</summary>

- **Why it belongs to this PR:** the guard is added by this diff; the PR body describes the empty-value skip only for the three `trim_text(true)` loops, not for `extract_dependencies`.

</details>

<sub>🤖 <b>mpiton-review</b> · product-logic · confidence 0.95</sub>

File: depsy-lsp/src/parsers/maven.rs (hunk 1 above)
Comment (mpiton) [inline depsy-lsp/src/parsers/maven.rs]: > [!NOTE]
> **🔵 Low · 🧪 Tests** — A value spanning two lines reports line_end < line_start and nothing asserts the invariant

`Span` is documented at `parsers/mod.rs:65-79` as a range within a single line, but `let (_, line_end) = offset_to_position(&offsets, ve)` throws away the end offset's line. Verified on this branch: `<version>1.0&amp;\n2</version>` gives `Span { line: 5, line_start: 9, line_end: 1 }` — an LSP range that ends before it starts. Not introduced here (`<version>1.0\n2</version>` without any entity gives the same span on this branch and on HEAD~1), but the PR rewrites this computation and makes the case reachable for entity-bearing values without pinning the invariant.

**Fix:** Where the span is built, drop it when `offset_to_position(&offsets, ve).0 != line` rather than emitting a reversed range. Then add maven to `validate_deps` in `depsy-lsp/tests/fuzz_regression.rs` — it already asserts `start <= end` and `end <= line_len`, but only for cargo, npm, php and python, so maven is the one span-producing parser it does not guard.

<details>
<summary>🔎 Details</summary>

- **Why it belongs to this PR:** the span now runs from the first fragment's `event_start` to the last fragment's position, which is what lets a single value straddle a line break.

</details>

<sub>🤖 <b>mpiton-review</b> · tests · confidence 0.9</sub>

File: depsy-lsp/src/utils.rs [hunk 5]
```diff
@@ -80,6 +82,44 @@ pub fn html_escape(s: &str) -> String {
     out
 }
 
+/// Appends the text an XML reference stands for to `out`.
+///
+/// quick-xml emits every `&…;` inside element content as a standalone
+/// [`Event::GeneralRef`](quick_xml::events::Event::GeneralRef) sandwiched
+/// between two `Event::Text` events, so a parser that assembles element text
+/// has to fold the reference back in itself. Numeric character references
```
Comment (mpiton) [inline depsy-lsp/src/utils.rs]: > [!NOTE]
> **🔵 Low · 📄 Docs** — push_xml_ref doc claims a reference is always sandwiched between two Text events

quick-xml emits no `Event::Text` on a side where the reference touches the element boundary: `<version>&amp;</version>` yields exactly one `GeneralRef` and zero `Text` events. This doc block exists to teach the next parser author the event model, and `extract_dependencies` already codes around the case it denies (`text_span.map_or(event_start, ..)` at line 348). The function is also the only helper in this file without a case in its own `mod tests` — `.github/workflows/coverage.yml` runs tarpaulin with the default `--run-types Tests`, which skips the doctest, so the hex form and the malformed-numeric fallback the doc advertises are unexercised in coverage.

**Fix:** Say the reference sits between the surrounding `Event::Text` runs, with no `Text` event on a side where it starts or ends the content. Add a `#[test] fn test_push_xml_ref()` covering `#x2D` → `-`, an out-of-range `#xD800` → `&#xD800;`, `lt`/`gt`/`quot`/`apos`, and `nbsp` → `&nbsp;`.

<details>
<summary>🔎 Details</summary>

- **Why it belongs to this PR:** both the doc block and the function are added by this diff.

</details>

<sub>🤖 <b>mpiton-review</b> · documentation · confidence 0.9</sub>

File: depsy-lsp/src/parsers/maven.rs (hunk 1 above)
Comment (mpiton) [reply depsy-lsp/src/parsers/maven.rs]: > A comment or CDATA inside an element is swallowed into the value and its span

Fixed in `3f52acb`, though not by resetting `text` on `Comment`/`PI`. I tried that first and it drops the value outright for `<version>1.7.30<!-- pinned --></version>`: the comment is the last thing in the element, so nothing follows to rebuild the text and the dependency disappears. Per XML the character data of `1.<!--c-->7.30` is `1.7.30`, so joining the fragments is right and only the span was wrong.

Each fragment now checks whether it starts where the previous one ended (`extend_span`), and a gap sets `text_gapped`, which drops the span and with it the declaration. That covers comments, PIs and anything else quick-xml doesn't route to the content arms, without enumerating event types. CDATA is appended and extends the span in all four loops, as you suggested.

Tests: `test_parse_trailing_comment_keeps_version_and_span`, `test_parse_comment_splitting_version_is_skipped`, `test_parse_cdata_inside_version_is_kept`.

File: depsy-lsp/src/parsers/maven.rs (hunk 2 above)
Comment (mpiton) [reply depsy-lsp/src/parsers/maven.rs]: > `<exclusions>` children overwrite the dependency's own groupId and artifactId

Confirmed, fixed in `3f52acb`. Added a `depth` counter and `dep_depth`; the coordinate match runs only when `dep_depth + 1 == depth`, so an `<exclusion>` child can't reach it. Dispatch moved onto `e.local_name()` and `current_tag` is gone, which also removes the coupling with the reset you pointed at.

`test_parse_exclusions_do_not_overwrite_coordinates` asserts `deps[0].name == "org.springframework:spring-core"` and that `name_span` lands on the dependency's own line, and there's an exclusions case in the new maven fuzz inputs.

File: depsy-lsp/src/parsers/maven.rs (hunk 3 above)
Comment (mpiton) [reply depsy-lsp/src/parsers/maven.rs]: > An unresolvable entity now emits a dependency whose version is the literal `&name;`

Confirmed, fixed in `3f52acb`. `push_xml_ref` now reports whether it resolved the reference, `extract_dependencies` records `text_unresolved`, and the `groupId`/`artifactId`/`version` arms are guarded on it, so the dependency gets skipped like before the PR. The verbatim fallback stays for free text (`<description>`, `<url>`, licence names) in the untouched `maven_central` loops, and the module doc's `Unsupported in this MVP` list records it. Test: `test_parse_unresolvable_entity_in_coordinate_is_skipped`.

One case this missed, caught by CodeRabbit on the same PR: `<properties>` had the same hole. `<ver>&custom;</ver>` with `<version>${ver}</version>` still reached Maven Central as `&custom;`, because the coordinate itself carries no reference and substitution runs after the check. `950579d` drops such a property instead, so `${ver}` stays unsubstituted like an undefined key.

File: depsy-lsp/src/parsers/maven.rs (hunk 4 above)
Comment (mpiton) [reply depsy-lsp/src/parsers/maven.rs:277]: > A whitespace-only <version> now drops the dependency instead of reporting it

Deliberate, and it stays. Same rule as a missing `<version>`: nothing to diagnose, nothing for the update quick-fix to write into. What was missing is the paper trail, so `test_parse_whitespace_only_version_is_skipped` now sits next to `test_parse_dependency_without_version_is_skipped` and the CHANGELOG entry names the case. Both in `3f52acb`.

File: depsy-lsp/src/parsers/maven.rs (hunk 1 above)
Comment (mpiton) [reply depsy-lsp/src/parsers/maven.rs]: > A value spanning two lines reports line_end < line_start and nothing asserts the invariant

Confirmed, fixed in `3f52acb`. `single_line_span` returns `None` when the end offset lands on a different line than the start. For `<version>` that drops the declaration; `name_span` falls back to the zero span. `test_parse_version_across_two_lines_is_skipped` pins it.

Maven is now in `validate_deps` in `fuzz_regression.rs` with 14 edge cases: entity splits, line breaks, mid-value comments, CDATA, exclusions, self-closing coordinates, a truncated document, control characters and non-ASCII.

File: depsy-lsp/src/utils.rs (hunk 5 above)
Comment (mpiton) [reply depsy-lsp/src/utils.rs]: > push_xml_ref doc claims a reference is always sandwiched between two Text events

Fixed in `3f52acb`. The doc now says the reference sits between the surrounding `Event::Text` runs, with no `Text` event on a side where it starts or ends the content, and uses `<version>&amp;</version>` as the example.

Three tests added in `utils::tests`, not a doctest, since tarpaulin's default `--run-types Tests` skips those: `test_push_xml_ref_resolves_numeric_refs` (`#45`, `#x2D`), `test_push_xml_ref_resolves_predefined_entities`, and `test_push_xml_ref_keeps_unresolvable_refs_verbatim` (`#xD800`, `nbsp`, malformed `#zz`). The function returns `bool` now, so the resolution outcome is pinned rather than just the output string.

File: depsy-lsp/src/parsers/maven.rs [hunk 6]
```diff
@@ -135,18 +148,24 @@ fn extract_properties(content: &str) -> HashMap<String, String> {
...
-                {
-                    // First occurrence wins to avoid overwriting project.version
-                    // with a nested <dependency><version>.
-                    out.insert(key.clone(), e.to_string());
-                }
+            Ok(Event::Text(e)) if current_key.is_some() => text.push_str(&e),
+            // CDATA carries the value literally, entity references included.
+            Ok(Event::CData(e)) if current_key.is_some() => text.push_str(&e),
+            Ok(Event::GeneralRef(e)) if current_key.is_some() => {
+                push_xml_ref(&mut text, &e);
```
[the remark this answers is not in the document]
Comment (mpiton) [reply depsy-lsp/src/parsers/maven.rs]: @coderabbitai

> Line 155 discards the `false` result from `push_xml_ref`. A property such as `<ver>&custom;</ver>` can then resolve `${ver}` to `&custom;` after the dependency coordinate has passed its local `text_unresolved` check.

Valid. Reproduced before the fix: `<ver>&custom;</ver>` with `<version>${ver}</version>` gave `version = "${lib.version}"`, `resolved_version = Some("&custom;")`, so `effective_version()` handed `&custom;` to the Maven Central lookup.

Fixed in `950579d`, but by dropping the property rather than threading its status through substitution. `extract_properties` exists only to feed `substitute` (one caller, `maven.rs:71`), so a value we can't read as written has nothing to contribute. Dropping it leaves `${ver}` unsubstituted and reuses the undefined-key path that already exists, instead of a parallel map plus a rejection rule at the emit site. Post-hoc scanning of the substituted value wasn't an option either, since a legitimately resolved `&amp;` produces the same `&` character.

`test_property_with_unknown_entity_is_dropped` covers the map, `test_unknown_entity_in_property_does_not_reach_the_version` covers the end-to-end path.

File: depsy-lsp/src/parsers/maven.rs [hunk 7]
```diff
@@ -251,30 +345,13 @@ fn extract_dependencies(content: &str, properties: &HashMap<String, String>) ->
...
-                                    let (line, line_start) = offset_to_position(&offsets, s);
-                                    let (_, line_end) = offset_to_position(&offsets, e_);
-                                    Span {
-                                        line,
-                                        line_start,
-                                        line_end,
-                                    }
-                                }
-                                None => Span {
+                            let name_span = artifact_span
```
[the remark this answers is not in the document]
Comment (mpiton) [reply depsy-lsp/src/parsers/maven.rs]: @cubic-dev-ai

> When an `<artifactId>` is split across lines or interrupted by a comment, `single_line_span` returns `None` but this fallback still emits a zero-length `name_span`

Valid. Reproduced before the fix: `<artifactId>my-\nlib</artifactId>` on line 5 gave `name_span = Span { line: 0, line_start: 0, line_end: 0 }`, so hover and the document link sat on `<project>` instead of the declaration. Same for `<artifactId>my<!-- x -->lib</artifactId>`, where the gap detection already nulls the span.

Fixed in `8a6dd54`. The name span joins the emit guard as `&& let Some(name_span) = artifact_span.and_then(...)`, so the dependency is skipped exactly like a `<version>` with no single-line range, and the zero-span fallback is gone. `test_parse_artifact_id_without_usable_span_is_skipped` covers both shapes.

File: depsy-lsp/src/registries/maven_central.rs [hunk 8]
```diff
@@ -152,41 +153,56 @@ pub(crate) fn parse_metadata_xml(
...
+                text.clear();
+                stack.push(e.name().as_ref().to_string());
             }
-            Ok(Event::Text(e)) => {
-                let text = e.to_string();
+            Ok(Event::Text(e)) => text.push_str(&e),
+            // CDATA carries the value literally, entity references included.
+            Ok(Event::CData(e)) => text.push_str(&e),
+            Ok(Event::GeneralRef(e)) => {
+                push_xml_ref(&mut text, &e);
```
[the remark this answers is not in the document]
Comment (mpiton) [reply depsy-lsp/src/registries/maven_central.rs]: @cubic-dev-ai

> When a metadata `<release>` or `<version>` contains an unresolved entity, this call stores the literal `&name;` and `get_version_info` interpolates it into the artifact URL.

Valid, and it's the third instance of the same hole after the dependency coordinates and the property map. Every value `parse_metadata_xml` keeps is a version, and `maven_central.rs:101` builds `{base}/{group_path}/{artifact}/{v}/{artifact}-{v}.pom` from it, so `&name;` became a request and got reported as the latest version. `maven-metadata.xml` is remote input, which is the reason to be strict about it.

Fixed in `8a6dd54` with the same pattern as the other two: `unresolved |= !push_xml_ref(...)`, and the `End` arm skips the value. `test_parse_metadata_xml_drops_unresolvable_entity` checks that `<release>` is dropped and that a sibling `<version>1.0</version>` still gets collected.

`parse_pom_metadata` in the same file is untouched on purpose. It only collects `<description>`, `<url>`, the SCM URL and licence names, which are free text, and `<description>Fast &amp; small</description>` is the case the verbatim fallback exists for.
